File: 1001-2000/1090-largest-values-from-labels/1090-largest-values-from-labels.cpp
class Solution
{
public:
    int largestValsFromLabels(vector<int>& values, vector<int>& labels, int numWanted, int useLimit)
    {
        vector<pair<int,int>> list(values.size());
        for(int i=0;i<values.size();i++)
        {
            list[i]={values[i],labels[i]};
        }
        sort(list.begin(),list.end(),greater<pair<int,int>>());
        int count[20001]={0};
        int sum=0;
        for(auto& m:list)
        {
            if(count[m.second]<useLimit)
            {
                count[m.second]++;
                numWanted--;
                sum+=m.first;
            }
            if(numWanted==0) break;
        }
        return sum;
    }
};

File: 1001-2000/1090-largest-values-from-labels/1090-largest-values-from-labels.java
class Solution {
    public int largestValsFromLabels(int[] values, int[] labels, int numWanted, int useLimit) {
        PriorityQueue<state>pq=new PriorityQueue<>((a,b)->(b.val-a.val));
        int n=values.length;
        for(int i=0;i<n;i++){
            pq.add(new state(values[i],labels[i]));
        }
        int []arr=new int[(2*10000)+1];
        int count=1;
        int sum=0;
        while(!pq.isEmpty()){
            if(count>numWanted) break;
            state cur=pq.poll();
            if(arr[cur.label]+1>useLimit) continue;
            arr[cur.label]++;
            sum+=cur.val;
            count++;
        }
        return sum;
    }
}
class state{
    int val;
    int label;
    public state(int val,int label){
        this.val=val;
        this.label=label;
    }
}